File: android/core/src/main/kotlin/org/hdhropen/kit/playback/VTTParser.kt
package org.hdhropen.kit.playback

data class CaptionCue(
    val start: Double,
    val end: Double,
    val text: String
) {
    val id: String get() = "${start}_${end}_${text.hashCode()}"

    fun contains(time: Double): Boolean = time in start..end
}

data class ThumbnailCue(
    val start: Double,
    val end: Double,
    val x: Int,
    val y: Int,
    val width: Int,
    val height: Int,
    val imageUrl: String? = null
) {
    val id: String get() = "${start}_${end}_${x}_${y}"

    fun contains(time: Double): Boolean = time in start..end
}

object VTTParser {
    fun parseTime(timeString: String): Double? {
        val trimmed = timeString.trim()
        val parts = trimmed.split(":")
        if (parts.size != 2 && parts.size != 3) return null

        return try {
            if (parts.size == 2) {
                val minutes = parts[0].toDouble()
                val seconds = parts[1].replace(",", ".").toDouble()
                minutes * 60.0 + seconds
            } else {
                val hours = parts[0].toDouble()
                val minutes = parts[1].toDouble()
                val seconds = parts[2].replace(",", ".").toDouble()
                hours * 3600.0 + minutes * 60.0 + seconds
            }
        } catch (e: Exception) {
            null
        }
    }

    fun parseCaptions(vttString: String): List<CaptionCue> {
        val cues = mutableListOf<CaptionCue>()
        val lines = vttString.lines()
        var i = 0

        while (i < lines.size) {
            val line = lines[i].trim()
            if (line.contains("-->")) {
                val timeParts = line.split("-->")
                if (timeParts.size == 2) {
                    val startStr = timeParts[0].trim()
                    val endStr = timeParts[1].trim().split(" ")[0]
                    val start = parseTime(startStr)
                    val end = parseTime(endStr)
                    if (start != null && end != null) {
                        val textLines = mutableListOf<String>()
                        i++
                        while (i < lines.size && lines[i].trim().isNotEmpty()) {
                            textLines.add(lines[i].trim())
                            i++
                        }
                        val text = textLines.joinToString("\n")
                        if (text.isNotEmpty()) {
                            cues.add(CaptionCue(start, end, text))
                        }
                    }
                }
            }
            i++
        }
        return cues
    }

    fun parseThumbnailVtt(vttString: String): List<ThumbnailCue> {
        val cues = mutableListOf<ThumbnailCue>()
        val lines = vttString.lines()
        var i = 0

        while (i < lines.size) {
            val line = lines[i].trim()
            if (line.contains("-->")) {
                val timeParts = line.split("-->")
                if (timeParts.size == 2) {
                    val startStr = timeParts[0].trim()
                    val endStr = timeParts[1].trim().split(" ")[0]
                    val start = parseTime(startStr)
                    val end = parseTime(endStr)
                    if (start != null && end != null) {
                        i++
                        if (i < lines.size) {
                            val mediaLine = lines[i].trim()
                            val xywhIdx = mediaLine.indexOf("#xywh=")
                            if (xywhIdx != -1) {
                                val imagePart = mediaLine.substring(0, xywhIdx)
                                val coordsPart = mediaLine.substring(xywhIdx + "#xywh=".length)
                                val coords = coordsPart.split(",").mapNotNull { it.trim().toIntOrNull() }
                                if (coords.size == 4) {
                                    cues.add(
                                        ThumbnailCue(
                                            start = start,
                                            end = end,
                                            x = coords[0],
                                            y = coords[1],
                                            width = coords[2],
                                            height = coords[3],
                                            imageUrl = imagePart.ifEmpty { null }
                                        )
                                    )
                                }
                            }
                        }
                    }
                }
            }
            i++
        }
        return cues
    }
}

File: apple/HDHROpenKit/Sources/HDHROpenKit/Playback/VTTParser.swift
import Foundation
import CoreGraphics

public struct CaptionCue: Identifiable, Sendable, Hashable {
    public var id: String { "\(start)_\(end)_\(text)" }
    public let start: Double
    public let end: Double
    public let text: String

    public init(start: Double, end: Double, text: String) {
        self.start = start
        self.end = end
        self.text = text
    }

    public func contains(time: Double) -> Bool {
        time >= start && time <= end
    }
}

public struct ThumbnailCue: Identifiable, Sendable, Hashable {
    public var id: String { "\(start)_\(end)_\(x)_\(y)" }
    public let start: Double
    public let end: Double
    public let x: Int
    public let y: Int
    public let width: Int
    public let height: Int
    public let imageURL: String?

    public init(start: Double, end: Double, x: Int, y: Int, width: Int, height: Int, imageURL: String? = nil) {
        self.start = start
        self.end = end
        self.x = x
        self.y = y
        self.width = width
        self.height = height
        self.imageURL = imageURL
    }

    public var cropRect: CGRect {
        CGRect(x: x, y: y, width: width, height: height)
    }

    public func contains(time: Double) -> Bool {
        time >= start && time <= end
    }
}

public enum VTTParser {
    public static func parseTime(_ timeString: String) -> Double? {
        let parts = timeString.trimmingCharacters(in: .whitespaces).components(separatedBy: ":")
        guard parts.count == 2 || parts.count == 3 else { return nil }

        if parts.count == 2 {
            guard let minutes = Double(parts[0]),
                  let seconds = Double(parts[1].replacingOccurrences(of: ",", with: ".")) else {
                return nil
            }
            return minutes * 60.0 + seconds
        } else {
            guard let hours = Double(parts[0]),
                  let minutes = Double(parts[1]),
                  let seconds = Double(parts[2].replacingOccurrences(of: ",", with: ".")) else {
                return nil
            }
            return hours * 3600.0 + minutes * 60.0 + seconds
        }
    }

    public static func parseCaptions(from vttString: String) -> [CaptionCue] {
        var cues: [CaptionCue] = []
        let lines = vttString.components(separatedBy: .newlines)
        var i = 0

        while i < lines.count {
            let line = lines[i].trimmingCharacters(in: .whitespaces)
            if line.contains("-->") {
                let timeParts = line.components(separatedBy: "-->")
                if timeParts.count == 2 {
                    let startStr = timeParts[0].trimmingCharacters(in: .whitespaces)
                    let endStr = timeParts[1].trimmingCharacters(in: .whitespaces).components(separatedBy: " ")[0]
                    if let start = parseTime(startStr), let end = parseTime(endStr) {
                        var textLines: [String] = []
                        i += 1
                        while i < lines.count && !lines[i].trimmingCharacters(in: .whitespaces).isEmpty {
                            textLines.append(lines[i].trimmingCharacters(in: .whitespaces))
                            i += 1
                        }
                        let text = textLines.joined(separator: "\n")
                        if !text.isEmpty {
                            cues.append(CaptionCue(start: start, end: end, text: text))
                        }
                    }
                }
            }
            i += 1
        }
        return cues
    }

    public static func parseThumbnailVtt(from vttString: String) -> [ThumbnailCue] {
        var cues: [ThumbnailCue] = []
        let lines = vttString.components(separatedBy: .newlines)
        var i = 0

        while i < lines.count {
            let line = lines[i].trimmingCharacters(in: .whitespaces)
            if line.contains("-->") {
                let timeParts = line.components(separatedBy: "-->")
                if timeParts.count == 2 {
                    let startStr = timeParts[0].trimmingCharacters(in: .whitespaces)
                    let endStr = timeParts[1].trimmingCharacters(in: .whitespaces).components(separatedBy: " ")[0]
                    if let start = parseTime(startStr), let end = parseTime(endStr) {
                        i += 1
                        if i < lines.count {
                            let mediaLine = lines[i].trimmingCharacters(in: .whitespaces)
                            if let xywhRange = mediaLine.range(of: "#xywh=") {
                                let imagePart = String(mediaLine[..<xywhRange.lowerBound])
                                let coordsPart = String(mediaLine[xywhRange.upperBound...])
                                let coords = coordsPart.components(separatedBy: ",").compactMap { Int($0) }
                                if coords.count == 4 {
                                    cues.append(ThumbnailCue(
                                        start: start,
                                        end: end,
                                        x: coords[0],
                                        y: coords[1],
                                        width: coords[2],
                                        height: coords[3],
                                        imageURL: imagePart
                                    ))
                                }
                            }
                        }
                    }
                }
            }
            i += 1
        }
        return cues
    }
}
